src/types.rs

```diff
@@ -1,31 +1,115 @@
 //! MQTT-Specific Data Types
 //!
 //! This module provides wrapper methods and serde functionality for MQTT-specified data types.
-use crate::{properties::Property, varint::Varint, QoS};
+use crate::{
+    de::deserializer::MqttDeserializer, properties::Property, varint::Varint, ProtocolError, QoS,
+};
 use bit_field::BitField;
 use serde::ser::SerializeStruct;
-use serde::Serialize;
+use serde::{Deserialize, Serialize};
 
-/// A wrapper type for a number of MQTT `Property`s.
-///
-/// # Note
-/// This wrapper type is primarily used to support custom serialization.
-#[derive(Copy, Clone, Debug, PartialEq)]
-pub struct Properties<'a>(pub &'a [Property<'a>]);
+#[derive(Debug, PartialEq)]
+pub enum Properties<'a> {
```

Comment:
They have to be the same type because the properties have to exist in the `PubAck` message type, which is both transmitted (serialized) and received (deserialize).

If we wanted them to be separate, we would have to have something along the lines of (this is not perfect, just a demonstration):
```rust
pub struct PubAck {
    reason_code: u16,
    #[serde(skip_when_serializing)]
    maybe_property_blob: Option<DataBlock>,
    #[serde(skip_when_deserializing)]
    maybe_property_slice: Option<&'a [Property<'a>]>,
}
```

This would recur for some other message types as we add QoS::ExactlyOnce support

Reply:
Ah. I missed that.